src/containers/CreatePipelineRun/CreatePipelineRun.js

```diff
@@ -14,497 +14,427 @@ limitations under the License.
-    regexp: resourceNameRegExp
-  },
-  imageRegistryName: {
-    required: false,
-    regexp: noSpacesRegExp
-  },
-  imageRepoName: {
-    required: false,
-    regexp: noSpacesRegExp
-  },
-  helmPipeline: {
-    required: false
-  },
-  helmSecret: {
-    required: false,
-    regexp: resourceNameRegExp
+const parsePipelineInfo = (state, pipelineRef, namespace) => {
+  if (pipelineRef) {
+    let resourceSpecs;
+    let paramSpecs;
+    const pipeline = getPipeline(state, { name: pipelineRef, namespace });
+    if (pipeline) {
+      ({ resources: resourceSpecs, params: paramSpecs } = pipeline.spec);
+    }
+    const pipelineError = !pipeline;
+    return {
+      [RESOURCE_SPECS]: resourceSpecs,
+      [PARAM_SPECS]: paramSpecs,
+      [PIPELINE_ERROR]: pipelineError
+    };
   }
+  return {};
 };
 
-const initialState = {
-  namespace: {
-    value: '',
-    invalid: false
-  },
-  pipeline: {
-    value: '',
-    invalid: false
-  },
-  serviceAccount: {
-    value: '',
-    invalid: false
-  },
-  gitName: {
-    value: '',
-    invalid: false
-  },
-  gitRevision: {
-    value: '',
-    invalid: false
-  },
-  gitRepoURL: {
-    value: '',
-    invalid: false
-  },
-  imageName: {
-    value: '',
-    invalid: false
-  },
-  imageRegistryName: {
-    value: '',
-    invalid: false
-  },
-  imageRepoName: {
-    value: '',
-    invalid: false
-  },
-  helmPipeline: {
-    value: false,
-    invalid: false
-  },
-  helmSecret: {
-    value: '',
-    invalid: false
-  },
-  errorMessage: '',
-  submitValidationError: false
+const initialPipelineInfoState = () => {
+  return {
+    [NAMESPACE]: '',
+    [PIPELINE_REF]: '',
+    [RESOURCE_SPECS]: [],
+    [PARAM_SPECS]: [],
+    [PIPELINE_ERROR]: false
+  };
+};
+
+const initialParamsState = paramSpecs => {
+  if (!paramSpecs) {
+    return {};
+  }
+  const paramsReducer = (acc, param) => ({
+    ...acc,
+    [param.name]: param.default || ''
+  });
+  return paramSpecs.reduce(paramsReducer, {});
+};
+
+const initialResourcesState = resourceSpecs => {
+  if (!resourceSpecs) {
+    return {};
+  }
+  const resourcesReducer = (acc, resource) => ({
+    ...acc,
+    [resource.name]: ''
+  });
+  return resourceSpecs.reduce(resourcesReducer, {});
 };
 
 class CreatePipelineRun extends React.Component {
   constructor(props) {
     super(props);
 
-    this.state = initialState;
-
-    this.onChange = this.onChange.bind(this);
-    this.onNamespaceChange = this.onNamespaceChange.bind(this);
-    this.onInputChange = this.onInputChange.bind(this);
-    this.onHelmToggleChange = this.onHelmToggleChange.bind(this);
-    this.onSubmit = this.onSubmit.bind(this);
-    this.onClose = this.onClose.bind(this);
-    this.reset = this.reset.bind(this);
-    this.getPipeline = this.getPipeline.bind(this);
     this.checkFormValidation = this.checkFormValidation.bind(this);
-    this.checkValidation = this.checkValidation.bind(this);
-  }
+    this.getPipelineInfo = this.getPipelineInfo.bind(this);
```

Comment:
it would be nice to remove these calls to `bind`, but I know there are some issues with tests when trying to spy on class properties so ok with leaving them if needed

Reply:
I can probably remove some of them. I always put `.bind` just to be safe, but I probably don't need to bind all of the functions? @AlanGreene can you please explain when I do and don't need to use `bind`?

Reply:
If the function is being called as a result of a user-event (e.g. click handler) you'll need to ensure it's bound to the component scope as by default it would be bound to the DOM node that was the target of the event (i.e. `this` would be a reference to the DOM node in the callback).

If the function is being called exclusively from other contexts where `this` is already the component, there's no need to bind.

Arrow functions (`() => { ... }`) (and class properties) and `bind` are just a few of the approaches to control scope in JS. You can read a little more about the background / reasoning for arrow functions here https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#No_separate_this and then follow the rabbit hole as deep as you like on JS scope in general 😉

Reply:
Ok, I think I've minimized the number of `bind`s, and it will still work 😄